datumaro/plugins/data_formats/datumaro/exporter.py

```diff
@@ -64,68 +65,89 @@ def items(self):
     def is_empty(self):
         return not self.items
 
-    def add_item(self, item: DatasetItem):
-        annotations = []
-        item_desc = {
-            "id": item.id,
-            "annotations": annotations,
-        }
-
-        if item.attributes:
-            item_desc["attr"] = item.attributes
-
-        if isinstance(item.media, Image):
+    @contextmanager
+    def save_media(self, item: DatasetItem):
+        """Implicitly change the media path and save it if save_media=True.
+        When done, revert it's path as before.
+        """
+        if item.media is None:
+            yield
+        elif isinstance(item.media, Image):
             image = item.media_as(Image)
             path = image.path
+
             if self._context._save_media:
-                path = self._context._make_image_filename(item)
-                self._context._save_image(
-                    item, osp.join(self._context._images_dir, item.subset, path)
+                # Temporarily update image path and save it.
+                image._path = osp.join(
+                    self._context._images_dir, item.subset, self._context._make_image_filename(item)
                 )
```

Comment:
is `_path` private variable of `MediaElement`? how about using `setter` and `getter` for path?

Reply:
Yes `path` was designed to be immutable but in this case, we have to make it mutable.
I think that the reason that we use `setter` in object oriented programming is when some member variables are dependent on the other variables. For example,

```python
class Whether:
def __init__(self, celsius: int):
     self._celsius = celsius
     self._text = "hot" if celsius > 30 else "warm"

@setter
def celsius(self, celsius: int):
     self._celsius = celsius
     self._text = "hot" if celsius > 30 else "warm"
```

You can see that `text` is dependent on `celsius` in this example. So, I decided to modify `_path` rather than creating the setter of it because it just increases the complexity of the code like
```python
@setter
def path(self, path: str):
    self._path = path
```

Reply:
okay, thanks.